lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp

```diff
@@ -0,0 +1,249 @@
+
+  virtual lldb::ValueObjectSP GetChildAtIndex(size_t idx);
+
+  virtual bool Update();
+
+  virtual bool MightHaveChildren();
+
+  virtual size_t GetIndexOfChildWithName(ConstString name);
+
+  virtual ~UnsafeBufferPointerSyntheticFrontEnd() = default;
+
+private:
+  ExecutionContextRef m_exe_ctx_ref;
+  uint8_t m_ptr_size;
+  lldb::ByteOrder m_order;
+
+  SwiftUnsafeBufferPointer m_unsafe_ptr;
+  size_t m_element_stride;
+  DataBufferSP m_buffer_sp;
+  std::vector<ValueObjectSP> m_children;
+};
+} // namespace swift
+} // namespace formatters
+} // namespace lldb_private
+
+lldb_private::formatters::swift::UnsafeBufferPointerSyntheticFrontEnd::
+    UnsafeBufferPointerSyntheticFrontEnd(lldb::ValueObjectSP valobj_sp)
+    : SyntheticChildrenFrontEnd(*valobj_sp.get()),
+      m_unsafe_ptr(*valobj_sp.get()) {
+
+  ProcessSP process_sp = valobj_sp->GetExecutionContextRef().GetProcessSP();
+  if (!process_sp)
+    return;
+
+  if (valobj_sp)
+    Update();
+
+  if (!m_unsafe_ptr.IsValid())
+    return;
+}
+
+size_t lldb_private::formatters::swift::UnsafeBufferPointerSyntheticFrontEnd::
+    CalculateNumChildren() {
+  return m_unsafe_ptr.GetCount();
+}
+
+lldb::ValueObjectSP lldb_private::formatters::swift::
+    UnsafeBufferPointerSyntheticFrontEnd::GetChildAtIndex(size_t idx) {
+  if (!m_unsafe_ptr.IsValid() && idx > m_unsafe_ptr.GetCount())
+    return ValueObjectSP();
+  const size_t num_children = CalculateNumChildren();
+
+  if (idx >= num_children || idx >= m_children.size())
+    return lldb::ValueObjectSP();
+
+  return m_children[idx];
+}
+
+bool lldb_private::formatters::swift::UnsafeBufferPointerSyntheticFrontEnd::
+    Update() {
+  m_children.clear();
+  ValueObjectSP valobj_sp = m_backend.GetSP();
+  m_ptr_size = 0;
+  if (!valobj_sp)
+    return false;
+  m_exe_ctx_ref = valobj_sp->GetExecutionContextRef();
+
+  lldb::ProcessSP process_sp(valobj_sp->GetProcessSP());
+  if (!process_sp)
+    return false;
+  m_ptr_size = process_sp->GetAddressByteSize();
+  m_order = process_sp->GetByteOrder();
+  if (!m_unsafe_ptr.Update())
+    return false;
+
+  const addr_t start_addr = m_unsafe_ptr.GetStartAddress();
+  const size_t num_children = CalculateNumChildren();
+  const CompilerType element_type = m_unsafe_ptr.GetElementType();
+
+  auto stride = element_type.GetByteStride(process_sp.get());
+  if (!stride)
+    return false;
+
+  m_element_stride = *stride;
+
+  if (m_children.empty()) {
+    size_t buffer_size = num_children * m_element_stride;
+    m_buffer_sp.reset(new DataBufferHeap(buffer_size, 0));
+
+    Status error;
+    size_t read_bytes = process_sp->ReadMemory(
+        start_addr, m_buffer_sp->GetBytes(), buffer_size, error);
+
+    if (!read_bytes || error.Fail())
+      return false;
+
+    for (size_t i = 0; i < num_children; i++) {
+      DataBufferSP buffer_sp(new DataBufferHeap(
+          m_buffer_sp->GetBytes() + i * m_element_stride,
+          m_element_stride)); // Create a buffer from contiguous memory
```

Comment:
Do not create a new DataBufferHeap for each child. You can create a DataExtractor from the main DataBuffer with an offset.